[tagger] pre-hash tags

vickenty opened the pull request:
### What does this PR do?

Store pre-computed hashes for the tags in the tagger, to avoid re-hashing the same strings for every metric sample.
Use local tagstore for the replay tagger, so it benefits from the same optimization.

### Motivation

Reduce CPU usage when computing context key. Tags from origin detection do not change frequently, so we save a significant amount of time by pre-computing hash values for tags that are mostly static. This reduces cpu load by roughly 13% in the benchmarks, although benefit varies depending on cardinality level used.

### Additional Notes

This PR contains two logically separate changes: optimization for the local tagstore, and a refactor that unifies tagstore for local and replay taggers. 

### Describe how to test your changes

Run the agent with docker or kubernetes, and origin detection and payload logging on:
```
dogstatsd_tag_cardinality: orchestrator
dogstatsd_socket: /tmp/dsd.sock
log_payloads: true
log_level: debug
```

Send some metrics with tags from a docker container.

Check that the metric is present in the logged payload with the correct tags.

Make a traffic recording using `agent dogstatsd-capture`.

Stop the client container.

Restart the agent.

Replay the traffic using `agent dogstatsd-replay`.

Check that the metric is present in the logged payload with the correct tags.

Do an benchmark to double-check the performance impact.

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

vickenty committed:
[ckey] move tag hashing to TagsBuilder

Hash tags as they are added to the TagsBuilder, and use these hashes
when generating context key.

Side benefit, SortUniq() for TagsBuilder() no longer allocates anything.

vickenty committed:
[tagger/local] cache hash values for tags

Keep tags in the local tagger pre-hashed, so that we don't have to hash
the same strings over and over when generating context key.

vickenty committed:
[tagger/local] extract tagstore package

vickenty committed:
[tagger/local] cleanup after tagstore move

vickenty committed:
[tagger/replay] use same tagstore as tagger/local

vickenty committed:
Add release note

vickenty committed:
[metrics] move EnrichTags into tagger

This code is shared between metrics, events and service checks.

vickenty committed:
[tagger] split off hashing out of TagsBuilder

Split tag hashing function out of TagsBuilder into a new type,
HashingTagsBuilder. This avoids computing tag hashes for events and
service checks.

vickenty committed:
Merge remote-tracking branch 'origin/main' into nohash-clean

vickenty committed:
[ckey] fix tags deduplication after merging main

djmitche commented on pkg/aggregator/ckey/key.go:
```diff
@@ -91,11 +91,11 @@ func (g *KeyGenerator) Generate(name, hostname string, tagsBuf *util.TagsBuilder
 		tagsBuf.SortUniq()
-		for _, tag := range tagsBuf.Get() {
-			h := murmur3.StringSum64(tag)
+		for _, h := range tagsBuf.Hashes() {
```
```suggestion
		for _, h := range hashes {
```

vickenty replied:
In this case no, as SortUniq would truncate inner slice. I moved local slice copies into the `if` branches to avoid confusion.

djmitche commented on pkg/tagger/local/tagger.go:
```diff
@@ -277,7 +276,6 @@ func (t *Tagger) getTags(entity string, cardinality collectors.TagCardinality) (
 
 	// Else, partial cache miss, query missing data
 	// TODO: get logging on that to make sure we should optimize
```
I'm not sure what this comment means -- is it still relevant?

djmitche commented on pkg/util/tags_builder.go:
```diff
@@ -86,16 +92,25 @@ func (tb *TagsBuilder) Truncate(len int) {
 func (tb *TagsBuilder) Get() []string {
+	if tb == nil {
+		return nil
+	}
```
I think it's unusual for methods to support a nil receiver -- perhaps a doc comment on the struct or each method would be useful?

djmitche commented on pkg/util/tags_builder.go:
```diff
@@ -105,6 +120,14 @@ func (tb *TagsBuilder) Less(i, j int) bool {
+func (tb *TagsBuilder) Slice(i, j int) *TagsBuilder {
+	return &TagsBuilder{
+		data: tb.data[i:j],
+		hash: tb.hash[i:j],
```
This looks pretty scary -- that slice will share the same underlying array, so any `slice.Append(..)` will re-use storage slots that the parent TagsBuilder is also using.

```go
tb1 := NewTagsBuilder(..) // with 15 items
tb2 := tb1.Slice(0, 10)
tb2.Append("foo")
assert.Equal(t, "foo", tb1.Get()[10])
```
Maybe it would be better for all tagsets to be immutable?  Or to have an immutability flag that is set to true in the TagsBuilder returned from Slice?

djmitche reviewed (commented):
These are some really nice refactors, and a good improvement too!

I'm not sure what "13% millicores" mean.  If the new CPU time is 13% smaller than the old, that's great!  If the total millicores used by the agent in production is 13% smaller, that's astonishing.

vickenty committed:
[tagger] use immutable hashed tags container

Split hashed tags builder into mutable (HashingTagsBuilder) and immutable
(HashedTags) variants, to prevent accidental modification of tags stored
in the taggers.

vickenty committed:
[tagger] move nil out to caller sites

Instead of checking for nil receiver inside a method, check at the
caller site and skip the call if the value is nil.

vickenty committed:
[ckey] small refactor

Keep slice copies local to each if arm that uses them.

djmitche commented on pkg/aggregator/ckey/key.go:
```diff
@@ -75,7 +75,7 @@ type KeyGenerator struct {
 
 // Generate returns the ContextKey hash for the given parameters.
 // tagsBuf is re-arranged in place and truncated to only contain unique tags.
```
Instead of modifying a value in-place, would it make sense to move the sort-and-dedup functionality into HashingTagsBuilder?  Then this method (Generate) would boil down to just constant ^ murmur(name) ^ murmur(hostname) ^ HashingTagsBuilder.DedupeAndHash().

vickenty replied:
Yes, with the latest changes it should be possible. But then it should probably be renamed and moved somewhere under aggregator, since it is becoming even more specific to context key generation. That would be a fairly large refactor, so maybe it would be better to do in a separate PR.

djmitche commented on pkg/aggregator/context_resolver.go:
```diff
@@ -38,7 +38,7 @@ func newContextResolver() *contextResolver {
 	return &contextResolver{
 		contextsByKey: make(map[ckey.ContextKey]*Context),
 		keyGenerator:  ckey.NewKeyGenerator(),
-		tagsBuffer:    util.NewTagsBuilder(),
```
Later in this file (not changed in this PR), `trackContext` creates a Context, which contains `Tags []string`.  Would it make sense to instead store a HashedTags instance there?

vickenty replied:
I don't think so, we only need the tags' hashes for the context key. Once the tags are stored in the context, we will basically not look at them again until we serialize metrics for the intake.

djmitche commented on pkg/util/tags_builder.go:
```diff
@@ -44,17 +61,171 @@ func (tb *TagsBuilder) Reset() {
+
 // Copy makes a copy of the internal slice
-func (tb *TagsBuilder) Copy() []string {
+func (tb *HashingTagsBuilder) Copy() []string {
```
It might make sense to have a `CopyAsHashedTags` that takes a "snapshot" of the builder as an immutable value.  This might be useful in `context_resolver.go` if my comment on that file makes sense.

It might also make sense to have a `HashedTags()` method that "converts" a `HashingTagsBuilder` into a `HashedTags` instance (invalidating or resetting the builder), although I don't actually see anywhere that conversion is useful.

djmitche reviewed (commented):
This looks great, and I think I've stared at it long enough to not see any issues around slice aliasing.  Hopefully the comments below make sense, and don't lead to too much change..

vickenty committed:
[util/tags_builder] make better use of embedding

Keep common methods on the base type; subtype specific methods can be
provided directly by the subtype.

vickenty committed:
[util/tags_builder] restore tests for TagsBuilder

vickenty commented on pkg/util/tags_builder_test.go:
```diff
@@ -12,20 +12,20 @@ import (
 )
 
 func TestNewBuilder(t *testing.T) {
-	tb := NewTagsBuilder()
```
Yes! Done.

vickenty commented on pkg/util/tags_builder.go:
```diff
@@ -44,17 +61,171 @@ func (tb *TagsBuilder) Reset() {
+	return HashedTags{newHashedTagsFromSlice(tags)}
+}
+
+// Slice returns a shared sub-slice of tags from t.
```
This method exists so we can read and modify the slice in ckey.Generate. Same is partially true for `Get()` as well.

vickenty replied:
> It should not be modified.
It can not be (directly), HashedTags has no mutators. 

It can still be invalidated (by mistake) by mutating slices returned from `Get` or `Hashes`, so not sure we can promise it to remain valid either. :/

vickenty commented on pkg/tagger/types/types.go:
```diff
@@ -77,3 +78,9 @@ type EntityEvent struct {
 	Entity    Entity
 }
+
+// TagsBuilder represents a consumer for tags stored in the tagger.
```
In a broad sense it is similar to producer/consumer - we give an object to the tagger, and it calls a method on it to feed it the data. As for the name, yes, it could. But this interface replaces a pre-existing type, and I think it is helpful to keep the same name, to preserve knowledge that people have about the code.

djmitche replied:
Hm, I'm not sure I agree with that reasoning, as it biases the code for understanding by people who already understand it, at the cost of everyone else.. But, OK :)

vickenty committed:
[util/tags_builder] remove duplicate methods

Get and Hashes are provided by the base type already.

vickenty committed:
Fix comment

vickenty committed:
Add comment about tags layout in the tagstore

vickenty committed:
Sync doc comments for Subscribe

subscriber.Subscribe had the most complete one, copy it to other locations.

vickenty committed:
Fix typo

Co-authored-by: Dustin J. Mitchell <dustin.mitchell@datadoghq.com>

vickenty committed:
[tagger/tagstore] update method name

vickenty commented on pkg/util/tags_builder.go:
```diff
@@ -44,17 +61,171 @@ func (tb *TagsBuilder) Reset() {
+	}
+}
+
+// NewHashingTagsBuilderFromSlice return a new TagsBuilder with the input slice for
+// it's internal buffer.
```
Not sure it makes sense to talk about ownership in Go, where everything is shared. Simply creating a HashingTagsBuilder with a slice doesn't really modify or invalidate caller's copy of the slice.

djmitche replied:
True, but if they reuse that slice, or even if the backing array for that slice is reused for other slices, that could cause difficult-to-debug problems.  I'm not sure how to describe that in go-ish terms.

vickenty replied:
Good point. This function was mostly used in tags, so I just changed to copy the slice instead.

vickenty commented on pkg/util/tags_builder.go:
```diff
@@ -44,17 +61,171 @@ func (tb *TagsBuilder) Reset() {
+	tb.hash = tb.hash[0:len]
 }
 
 // Get returns the internal slice
```
We do modify the returned slice in Generate (same true for Hashes).

djmitche replied:
That's partly why I was suggesting making Generate a method of HashingTagBuilder -- it would avoid having HashingTagBuilder's internal state mutated by something external.

vickenty commented on pkg/tagger/tagstore/tagstore.go:
```diff
@@ -338,30 +360,66 @@ type tagPriority struct {
-		HighCardinalityTags:         e.cachedAll[len(e.cachedOrchestrator):],
-		OrchestratorCardinalityTags: e.cachedOrchestrator[len(e.cachedLow):],
-		LowCardinalityTags:          e.cachedLow,
+		HighCardinalityTags:         cachedAll[len(cachedOrchestrator):],
+		OrchestratorCardinalityTags: cachedOrchestrator[len(cachedLow):],
+		LowCardinalityTags:          cachedLow,
```
It skips orchestrator tags instead of selecting them; I added some comments above (https://github.com/DataDog/datadog-agent/commit/c43482705bc846084f779a150c66e502d73d1d0e)

vickenty committed:
gofmt

vickenty committed:
[tags_builder] improve ctor

Remove NewHashingTagsBuilderFromSlice as a potential footgun, and
replace with a constructor that copies tags instead of re-using the
slice, NewHashingTagsBuilderWithTags.

Replace the only use outside of tests with a persistent instance.

vickenty committed:
[tagger] fix data race

Copy HashedTags out of entityTags instead of returning a pointer to a
field.

djmitche reviewed (approved):
All of the changes look great.  Thanks for putting them in individual commits!
